File: app/src/test/java/io/github/tobyhs/weatherweight/test/ForecastResultSetFactory.java
package io.github.tobyhs.weatherweight.test;

import java.util.Arrays;
import java.util.List;

import org.threeten.bp.LocalDate;
import org.threeten.bp.ZonedDateTime;

import io.github.tobyhs.weatherweight.data.model.DailyForecast;
import io.github.tobyhs.weatherweight.data.model.ForecastResultSet;

/**
 * This creates {@link ForecastResultSet} objects for tests
 */
public class ForecastResultSetFactory {
    /**
     * @return a valid {@link ForecastResultSet} for testing
     */
    public static ForecastResultSet create() {
        return ForecastResultSet.builder()
                .setLocation("Oakland, CA, US")
                .setPublicationTime(ZonedDateTime.parse("2019-02-01T12:00:00Z"))
                .setForecasts(createForecasts())
                .build();
    }

    /**
     * @return a list of daily forecasts for testing
     */
    public static List<DailyForecast> createForecasts() {
        return Arrays.asList(
                DailyForecast.builder().setDate(LocalDate.parse("2019-02-01"))
                        .setLow(60).setHigh(65).setText("Cloudy").build(),
                DailyForecast.builder().setDate(LocalDate.parse("2019-02-02"))
                        .setLow(58).setHigh(64).setText("Showers").build()
        );
    }
}

File: app/src/test/java/io/github/tobyhs/weatherweight/test/ForecastResultSetFactory.kt
package io.github.tobyhs.weatherweight.test

import io.github.tobyhs.weatherweight.data.model.DailyForecast
import io.github.tobyhs.weatherweight.data.model.ForecastResultSet
import io.github.tobyhs.weatherweight.data.model.HourlyForecast

import java.time.LocalDate
import java.time.ZonedDateTime

/**
 * This creates [ForecastResultSet] objects for tests
 */
object ForecastResultSetFactory {
    /**
     * @return a valid [ForecastResultSet] for testing
     */
    fun create(): ForecastResultSet = ForecastResultSet(
        location = "Oakland, CA, US",
        publicationTime = ZonedDateTime.parse("2019-02-01T12:00:00Z"),
        dailyForecasts = createDailyForecasts(),
        hourlyForecasts = createHourlyForecasts(),
    )

    /**
     * @return a list of daily forecasts for testing
     */
    fun createDailyForecasts(): List<DailyForecast> = listOf(
        DailyForecast(
            date = LocalDate.parse("2019-02-01"),
            low = 60,
            high = 65,
            text = "Cloudy",
            precipitationProbability = 10,
        ),
        DailyForecast(
            date = LocalDate.parse("2019-02-02"),
            low = 58,
            high = 64,
            text = "Showers",
            precipitationProbability = 70,
        ),
    )

    /**
     * @return a list of hourly forecasts for testing
     */
    fun createHourlyForecasts(): List<HourlyForecast> = listOf(
        HourlyForecast(
            dateTime = ZonedDateTime.parse("2019-02-01T11:00:00Z"),
            text = "Mostly clear",
            temperature = 63,
            precipitationProbability = 5,
        ),
        HourlyForecast(
            dateTime = ZonedDateTime.parse("2019-02-01T12:00:00Z"),
            text = "Cloudy",
            temperature = 67,
            precipitationProbability = 10,
        )
    )
}
